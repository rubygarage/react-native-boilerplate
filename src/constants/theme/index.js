export default {
  colors: {
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
  },
  sizes: {
  },
};
