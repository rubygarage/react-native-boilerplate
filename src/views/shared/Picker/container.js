import React, { Component } from 'react';
import { find, propEq, prop } from 'ramda';
import { withTheme } from 'styled-components/native';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';
import PropTypes from 'prop-types';

import PickerComponent from './component';

class Picker extends Component {
  state = { isOpen: false }

  get label() {
    const { selectedValue, options, intl: { formatMessage } } = this.props;
    const option = find(propEq('value', selectedValue))(options);
    return prop('label', option) || formatMessage({ id: 'shared.na' });
  }

  handlePickerToggle = () => {
    this.setState((prevState) => ({ isOpen: !prevState.isOpen }));
  }

  handleValueChange = (value) => {
    const { onValueChange } = this.props;
    onValueChange(value);
    this.handlePickerToggle();
  }

  render() {
    return (
      <PickerComponent
        {...this.props}
        {...this.state}
        label={this.label}
        onPickerToggle={this.handlePickerToggle}
        onValueChange={this.handleValueChange}
      />
    );
  }
}

Picker.propTypes = {
  onValueChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })).isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
  selectedValue: PropTypes.string,
};

Picker.defaultProps = {
  selectedValue: undefined,
};

const enhance = compose(
  withTheme,
  injectIntl,
);

export { Picker as PickerContainer };
export default enhance(Picker);
