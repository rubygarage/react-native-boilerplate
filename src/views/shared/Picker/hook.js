import { useState, useContext } from 'react';
import * as R from 'ramda';
import { ThemeContext } from 'styled-components';
import { useIntl } from 'react-intl';

const useContainer = (props) => {
  const {
    selectedValue,
    options,
    intl: { formatMessage },
    onValueChange,
  } = props;

  const theme = useContext(ThemeContext);
  const intl = useIntl();

  const [isOpen, setIsOpen] = useState(false);

  const detectLabel = () => {
    const option = R.find(R.propEq('value', selectedValue))(options);

    return R.prop('label', option) || formatMessage({ id: 'shared.na' });
  };

  const handlePickerToggle = () => {
    setIsOpen((prevIsOpen) => !prevIsOpen);
  };

  const handleValueChange = (value) => {
    onValueChange(value);

    handlePickerToggle();
  };

  return {
    theme,
    intl,
    isOpen,
    setIsOpen,
    detectLabel,
    handlePickerToggle,
    handleValueChange,
  };
};

export default useContainer;
