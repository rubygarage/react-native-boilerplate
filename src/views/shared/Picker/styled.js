import styled from 'styled-components/native';
import { Picker } from '@react-native-community/picker';

const PickerView = styled(Picker)`
  flex: 1;
`;

const PickerValueWrap = styled.View`
  flex-direction: row;
  align-items: center;
`;

const PickerValue = styled.Text`
  font-size: 14;
  line-height: 20;
  color: ${({ theme }) => theme.colors.black};
  margin-right: 8;
`;

export { PickerView, PickerValueWrap, PickerValue };
