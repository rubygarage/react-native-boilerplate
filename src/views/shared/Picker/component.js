import React from 'react';
import { TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';

import Icon from 'views/shared/Icon';
import SimplePicker from './SimplePicker';
import * as S from './styled';

const PickerComponent = (props) => {
  const { onPickerToggle, isOpen, label, theme } = props;
  return (isOpen
    ? <SimplePicker {...props} />
    : (
      <TouchableOpacity onPress={onPickerToggle}>
        <S.PickerValueWrap>
          <S.PickerValue>
            {label}
          </S.PickerValue>
          <Icon name="down-arrow" size={20} color={theme.colors.grey500} />
        </S.PickerValueWrap>
      </TouchableOpacity>
    ));
};

PickerComponent.propTypes = {
  onPickerToggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
};

export default PickerComponent;
