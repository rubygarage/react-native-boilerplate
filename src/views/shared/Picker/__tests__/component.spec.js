import React from 'react';
import { shallow } from 'enzyme';

import theme from 'utils/testHelpers/mockedTheme';
import Picker from '../component';

describe('Picker component', () => {
  const defaultProps = {
    onPickerToggle: jest.fn(),
    isOpen: false,
    label: 'value1',
    onValueChange: jest.fn(),
    selectedValue: 'value1',
    options: [
      { key: 'key0', label: 'label0', value: 'value0' },
      { key: 'key1', label: 'label1', value: 'value1' }],
    theme,
  };
  const component = shallow(<Picker {...defaultProps} />);

  it('renders correctly closed', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correctly opened', () => {
    component.setProps({ ...defaultProps, isOpen: true });
    expect(component).toMatchSnapshot();
  });
});
