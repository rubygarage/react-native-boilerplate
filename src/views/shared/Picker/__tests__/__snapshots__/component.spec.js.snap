// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker component renders correctly closed 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Styled(View)>
    <Styled(Text)>
      value1
    </Styled(Text)>
    <Icon
      allowFontScaling={false}
      color="[mocked grey500]"
      name="down-arrow"
      size={20}
    />
  </Styled(View)>
</TouchableOpacity>
`;

exports[`Picker component renders correctly opened 1`] = `
<SimplePicker
  isOpen={true}
  label="value1"
  onPickerToggle={[MockFunction]}
  onValueChange={[MockFunction]}
  options={
    Array [
      Object {
        "key": "key0",
        "label": "label0",
        "value": "value0",
      },
      Object {
        "key": "key1",
        "label": "label1",
        "value": "value1",
      },
    ]
  }
  selectedValue="value1"
  theme={
    Object {
      "colors": null,
      "fonts": null,
      "sizes": null,
      "type": "Mocked theme",
    }
  }
/>
`;
